core/src/test/java/hudson/model/RunTest.java

```diff
@@ -262,4 +266,50 @@ public void compareRunsFromDifferentParentsWithSameNumber() throws Exception {
         assertTrue(r1.compareTo(r2) != 0);
         assertTrue(treeSet.size() == 2);
     }
+
+    @Test
+    public void willTriggerLogToStartWithNextFullLine() throws Exception {
+        assertWriteLogToEquals(new String(new char[2]).replace("\0", SAMPLE_BUILD_OUTPUT) + "Finished: SUCCESS.\n", 2 * SAMPLE_BUILD_OUTPUT.length() + 10);
+    }
+
+    @Test
+    public void wontPushOffsetOnRenderingFromBeginning() throws Exception {
+        assertWriteLogToEquals(new String(new char[5]).replace("\0", SAMPLE_BUILD_OUTPUT) + "Finished: SUCCESS.\n", 0);
+    }
```

Comment:
Note that

```java
@Test
public void wontPushOffsetOnRenderingFromBeginningOfLine() throws Exception {
    assertWriteLogToEquals(new String(new char[3]).replace("\0", SAMPLE_BUILD_OUTPUT) + "Finished: SUCCESS.\n", 2 * SAMPLE_BUILD_OUTPUT.length());
}
```

fails: it prints 2 copies, not 3 like you might expect. This is because of a “fencepost error”: when the `offset` happens to point to the beginning of a line (just not the first line), that line is skipped.

If this method with `offset > 0` is only being used to skip 150Kb of text or whatever, the bug would hardly be noticeable.